Add misc.colored_clips to return random uniquely colored BlankClips

OrangeChannel opened the pull request:
This may or may not be useful at all. I used it to test my comparison functions.

Typehints passed mypy with no errors.

heicrd commented:
Might be a good idea to allow the user to provide an RNG seed for reproducibility

heicrd commented on lvsfunc/misc.py:
```diff
@@ -408,6 +410,53 @@ def upscale_chroma(n: int, f: vs.VideoFrame) -> vs.VideoNode:
+    if not (0 < max_hue <= 360):
+        raise ValueError("colored_clips: `max_hue` must be greater than 0 and less than 360 degrees")
+
+    blank_clip_args = dict(keep=1)
```
this needs to be `blank_clip_args: Dict[str, Any] = ...` to make mypy happy with vapoursynth typestubs

heicrd commented on lvsfunc/misc.py:
```diff
@@ -408,6 +410,53 @@ def upscale_chroma(n: int, f: vs.VideoFrame) -> vs.VideoNode:
+    for hue in hues:
+        hls_color_list.append(list(colorsys.hls_to_rgb(hue, .5, 1)))
+
+    rgb_color_list: List[List[int]] = [[int(f * 255) for f in color] for color in hls_color_list]
```
if this has the possibility of overflowing an 8bit color then just do `min(int(f * 255), 255)` here and remove lines 452-455
Edit: doesn't seem like this is a possibility so this can be left as-is and the other lines removed

heicrd commented on lvsfunc/misc.py:
```diff
@@ -408,6 +410,53 @@ def upscale_chroma(n: int, f: vs.VideoFrame) -> vs.VideoNode:
+    Returns a list of BlankClips with randomized unique colors. Useful maybe for comparison functions
+    or just for getting multiple uniquely colored BlankClips for testing purposes.
+
+    `max_hue` is the maximum hue that will be returned from the HSL color model.
```
This is kind of redundant with the param documentation below

heicrd commented on lvsfunc/misc.py:
```diff
@@ -408,6 +410,53 @@ def upscale_chroma(n: int, f: vs.VideoFrame) -> vs.VideoNode:
+
+    Written by Dave <orangechannel@pm.me>.
+
+    :param amount:  Number of `vapoursynth.VideoNode`s to return
```
you probably want \`\`vapoursynth.VideoNode\`\` for code tags, single backticks will render as italics

heicrd commented on lvsfunc/misc.py:
```diff
@@ -408,6 +410,53 @@ def upscale_chroma(n: int, f: vs.VideoFrame) -> vs.VideoNode:
 
+def colored_clips(amount: int, max_hue: int = 300, **kwargs: Any) -> List[vs.VideoNode]:
+    """
+    Returns a list of BlankClips with randomized unique colors. Useful maybe for comparison functions
```
This documentation seems to be inaccurate, as it produces evenly spaced hues in a randomized order

heicrd commented on lvsfunc/misc.py:
```diff
@@ -408,6 +410,53 @@ def upscale_chroma(n: int, f: vs.VideoFrame) -> vs.VideoNode:
+    hls_color_list: List[List[float]] = []
+
+    for hue in hues:
+        hls_color_list.append(list(colorsys.hls_to_rgb(hue, .5, 1)))
```
may be easier to read this as `hls_list: List[List[float]] = [list(colorsys.hls_to_rgb(h/360, 0.5, 1)) for h in hues]` instead of having it split across 4 lines

heicrd commented on lvsfunc/misc.py:
```diff
@@ -408,6 +410,53 @@ def upscale_chroma(n: int, f: vs.VideoFrame) -> vs.VideoNode:
+    hues: List[Union[float, int]] = [0]
+
+    for _ in range(amount - 2):
+        hues.append(hues[-1] + diff)
```
you could also do `hues = [x*(max_hue)/(amount-1) for x in range(amount-1)]` and follow up with `hues.append(max_hue)`

OrangeChannel commented:
Regarding the RNG seed, it would need to be a bit complex as the random functions don't take one by default, so you have initialize your own Random class with it, and I feel like it probably won't be used here by anyone. I instead added a `rand` parameter that would let the user return the same colors each time if `False`.

I also fixed up all the docstring stuff you and petzku brought up in Discord.

Passes mypy check again and return same result as the original commit with some basic vspreview testing.

OrangeChannel commented:
If you'd like, I can ff this branch to be level with the current master branch and squash down the commit into the first one after you check I addressed the feedback correctly @heicrd . We can then just commit this ontop I think.

heicrd commented:
>the random functions don't take one by default
Apparently you can do `random.Random(seed).shuffle(list)`
so something like
```python
shuffle = random.shuffle if seed is None else random.Random(seed).shuffle
shuffle(list)
```

OrangeChannel commented:
Oh that's actually pretty clean and simple, sounds good.

OrangeChannel commented:
Ready for merge now

FichteFoll commented on lvsfunc/misc.py:
```diff
@@ -415,6 +417,50 @@ def upscale_chroma(n: int, f: vs.VideoFrame) -> vs.VideoNode:
+
+    blank_clip_args: Dict[str, Any] = {'keep': 1, **kwargs}
+
+    hues: List[Union[float, int]] = [i * max_hue / (amount - 1) for i in range(amount - 1)]
```
`/` always returns a float, so there is no need for the union.

OrangeChannel replied:
That's because I append `max_hue` at the end and mypy doesn't like that since it's an int, also its kind of an artifact of originally starting the list with just `0`.

FichteFoll commented on lvsfunc/misc.py:
```diff
@@ -415,6 +417,50 @@ def upscale_chroma(n: int, f: vs.VideoFrame) -> vs.VideoNode:
+    Written by Dave <orangechannel@pm.me>.
+
+    :param amount:  Number of ``vapoursynth.VideoNode``s to return
+    :param max_hue: Maximum hue (0 < hue <= 360) in degrees to generate colors from (uses the HSL color model).
+                    Setting this higher than ``315`` will result in the clip colors looping back towards red and is not recommended for visually distinct colors.
+                    If the `amount` of clips is higher than the `max_hue` expect there to be identical or visually similar colored clips returned (Default: 300)
```
Does it make sense to have such a parameter in the first place? If you consider the hue plane to be cyclic from 0 to 360 (which it is), you just divide 360 into evenly spaced segments.

```py
>>> (lambda count: [360 / (i + 1) for i in range(count)])(6)
[360.0, 180.0, 120.0, 90.0, 72.0, 60.0]
```

OrangeChannel replied:
Try this with 16 clips though (which is what I did to test a mosaic compare). The end value is 337.5 which is basically red. I know by limiting it to 300 there's less space between the colors too which makes them *all* look more similar but I thought it'd be better to stop the colors from bleeding into the red section again. This lambda might make more sense though.
